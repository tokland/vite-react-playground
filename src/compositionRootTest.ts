import {
    CompositionRoot,
    Repositories,
    Services,
    getCompositionRoot,
    getAppCompositionRoot,
    getRepositories,
    getAppRepositories,
    getServices,
} from "./compositionRoot";
import { getProxy } from "./tests/AppProxySnapshots";

export type { CompositionRoot, Repositories, Services };

export function getProxiedRepositories() {
    const repositories = getAppRepositories();
    return getProxy(repositories, { type: { name: "Repositories", path: __filename } });
}

export async function getRepositoriesWithProxiedServices() {
    const services = getServices();
    const proxiedServices = await getProxy(services, {
        type: { name: "Services", path: __filename },
    });
    return getRepositories(proxiedServices);
}

export async function getCompositionRootWithProxiedRepos() {
    const proxiedRepositories = await getProxiedRepositories();
    return getCompositionRoot(proxiedRepositories);
}

export async function getProxiedCompositionRoot(): Promise<CompositionRoot> {
    const compositionRoot = getAppCompositionRoot();
    return getProxy(compositionRoot, { type: { name: "CompositionRoot", path: __filename } });
}
