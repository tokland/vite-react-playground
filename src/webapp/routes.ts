import { Route, createRouter, defineRoute, param } from "type-route";

export const {
    RouteProvider: RouteProvider,
    useRoute: useRoute,
    routes: routes,
} = createRouter({
    home: defineRoute("/"),
    counter: defineRoute({ id: param.path.string }, params => `/counter/${params.id}`),
});

export type Routes = Route<typeof routes>;
