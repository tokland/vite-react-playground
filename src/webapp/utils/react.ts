import React from "react";

export type Element = JSX.Element;

export function buildStyles<T extends Record<string, React.CSSProperties>>(styles: T): T {
    return styles;
}
