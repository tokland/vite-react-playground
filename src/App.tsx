import "./App.css";
import React from "react";
import { RouteProvider } from "./routes";
import Router from "./Router";
import { AppActions, AppState, useAppStore } from "./AppStore";
import { Counter } from "./domain/entities/Counter";
import { HashMap } from "./domain/entities/Generic/HashMap";

const initialState: AppState = {
    counters: HashMap.fromObject<string, Counter>({
        "1": Counter.create({ id: "1", value: 1 }),
        "2": Counter.create({ id: "2", value: 2 }),
    }),
};

function App() {
    const [AppProvider] = useAppStore(accessors => ({
        initialState: initialState,
        actions: new AppActions(accessors.set),
    }));

    return (
        <div className="App">
            <RouteProvider>
                <AppProvider>
                    <Router />
                </AppProvider>
            </RouteProvider>
        </div>
    );
}

export default React.memo(App);
