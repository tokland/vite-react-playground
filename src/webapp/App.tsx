import "./App.css";
import React from "react";
import { RouteProvider } from "./routes";
import Router from "./Router";
import { AppActions, AppState, useAppStore } from "./AppStore";
import { CompositionRoot } from "../compositionRoot";
import * as entities from "../domain/entities";

const initialState: AppState = {
    counters: entities.Counters.empty(),
};

function App(props: { compositionRoot: CompositionRoot }) {
    const { compositionRoot } = props;

    const [AppProvider] = useAppStore(accessors => ({
        initialState: initialState,
        actions: new AppActions(accessors.get, accessors.set, compositionRoot),
    }));

    return (
        <div className="App">
            <RouteProvider>
                <AppProvider>
                    <Router />
                </AppProvider>
            </RouteProvider>
        </div>
    );
}

export default React.memo(App);
