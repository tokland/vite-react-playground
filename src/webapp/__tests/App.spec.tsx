import { act, fireEvent, render, waitFor } from "@testing-library/react";

import App from "../App";
import { getProxiedCompositionRoot } from "../../compositionRootTest";
import { routes } from "../routes";

describe("App", () => {
    it("renders components", async () => {
        act(() => routes.counter({ id: "1" }).push());

        const compositionRoot = await getProxiedCompositionRoot();
        const view = render(<App compositionRoot={compositionRoot} />);

        await waitFor(() => expect(view.getByText("+1")));
        fireEvent.click(view.getByText("+1"));
        fireEvent.click(view.getByText("+1"));
        expect(view.asFragment()).toMatchSnapshot();
    });
});
