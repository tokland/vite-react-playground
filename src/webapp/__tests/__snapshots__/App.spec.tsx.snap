// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders components 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav>
      <a
        href="/"
      >
        Home
      </a>
      |
      <span>
        Counter 1
      </span>
      |
      <a
        href="/counter/2"
      >
        Counter 2
      </a>
    </nav>
    <div
      class="App"
    >
      <div
        class="card"
      >
        <button
          data-testid="increment-counter"
        >
          -1
        </button>
        <span
          data-testid="counter-value"
          style="padding: 10px;"
        >
          val=2
        </span>
        <button
          data-testid="decrement-counter"
        >
          +1
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
